data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetFinalizedBlockRoot.java

```diff
@@ -0,0 +1,101 @@
+import tech.pegasys.teku.infrastructure.restapi.endpoints.EndpointMetadata;
+import tech.pegasys.teku.infrastructure.restapi.endpoints.RestApiRequest;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.spec.datastructures.metadata.ObjectAndMetaData;
+
+public class GetFinalizedBlockRoot extends MigratingEndpointAdapter {
+
+  public static final String ROUTE = "/eth/v1/checkpoint/finalized_blocks/{slot}/root";
+  private final ChainDataProvider chainDataProvider;
+
+  private static final SerializableTypeDefinition<ObjectAndMetaData<Bytes32>> RESPONSE_TYPE =
+      SerializableTypeDefinition.<ObjectAndMetaData<Bytes32>>object()
+          .name("GetFinalizedBlockRootResponse")
+          .withField("data", ROOT_TYPE, ObjectAndMetaData::getData)
+          .withField(EXECUTION_OPTIMISTIC, BOOLEAN_TYPE, ObjectAndMetaData::isExecutionOptimistic)
+          .build();
+
+  public GetFinalizedBlockRoot(DataProvider dataProvider) {
+    this(dataProvider.getChainDataProvider());
+  }
+
+  public GetFinalizedBlockRoot(ChainDataProvider chainDataProvider) {
+    super(
+        EndpointMetadata.get(ROUTE)
+            .operationId("getFinalizedBlockRoot")
+            .summary("Get finalized block root")
+            .description(
+                "Retrieves hashTreeRoot of finalized BeaconBlock/BeaconBlockHeader.\n"
+                    + "Responds with 404 if block at a slot is either unavailable or not yet finalized.")
+            .tags(TAG_EXPERIMENTAL)
+            .pathParam(SLOT_PARAMETER)
+            .response(SC_OK, "Request successful", RESPONSE_TYPE)
+            .withNotFoundResponse()
+            .build());
+    this.chainDataProvider = chainDataProvider;
+  }
+
+  @Override
+  public void handle(@NotNull Context ctx) throws Exception {
+    adapt(ctx);
+  }
+
+  @Override
+  public void handleRequest(RestApiRequest request) throws JsonProcessingException {
+    final UInt64 slot = request.getPathParameter(SLOT_PARAMETER);
+    final SafeFuture<Optional<SignedBeaconBlock>> eventuallyFinalizedBlock =
+        chainDataProvider.getFinalizedBlockInEffectAtSlot(slot);
+    request.respondAsync(
+        eventuallyFinalizedBlock.thenApply(
+            maybeFinalizedBlock ->
+                maybeFinalizedBlock
+                    .map(
+                        finalizedBlock -> {
+                          if (finalizedBlock.getSlot().equals(slot)) {
+                            chainDataProvider
+                                .getBlockRoot(slot.toString())
+                                .thenApply(
+                                    maybeFinalizedBlockRoot ->
+                                        maybeFinalizedBlockRoot
+                                            .map(AsyncApiResponse::respondOk)
+                                            .orElse(AsyncApiResponse.respondNotFound()));
```

Comment:
if we have the whole block here, i don't see why we would re-fetch the root...

one possibility is we need to do more than 1 query to figure out the result, and we're trying to do some business logic at the api level?
my preference if we actually need 2 calls, is pass information to chainDataProvider, and it responds with `Optional<BlockRoot>`, and then the business logic is out of the api... 

Business logic should be ideally shared if we ever need it again, plus chainDataProvider is way easier to test than interfaces, but if we've got logic here then it's going to be complicated to prove it's what we want to do...

Reply:
looking more into hte ticket etc, i think once we have finalizedBlockInEffect (first future), we then respond with block.getRoot(), rather than doing another query, we have the block we want, we just need to get the data from it I think...

Reply:
Yep so I think typically we'd just have something like:

```
maybeFinalizedBlock 
    .filter(block -> block.getSlot().equals(slot))
    .map(SignedBeaconBlock::getRoot)
    .map(AsyncApiResponse::responseOk)
    .orElseGet(AsyncApiRessponse::respondNotFound)
```

As Paul says, probably makes sense to push this logic into `ChainDataProvider` as a new `getFinalizedBlockRoot` method.

Reply:
Done: moved the logic to the `ChainDataProvider` service